### ISSUE TITLE
Explanation of nested reference please

### ISSUE THREAD

**Issue by Tanoti** [sphinx-contrib/confluencebuilder#266]
I've got a large set of auto-generated docs and I am getting this error when trying to generate the confluence pages:

```
Exception occurred:
  File "c:\salt-master-fileset\venv\lib\site-packages\sphinxcontrib\confluencebuilder\translator.py", line 779, in visit_reference
    assert(not self._reference_context)
AssertionError
```

I can see from the code at the point that it is a catch for nested references but what would one of those look like? How can I track down the cause of this?

Thanks in advance!

**Comment by jdknight** [comment]
Interesting find. This extension's implementation assumes that nested references would never happen, but clearly the documentation set being processed is using such a capability provided by Sphinx.

If approaching this at a user level, one could adjust the invoke Sphinx process to include verbose messages (`-v`) which in turn should provide better output to which document is causing the issue. After that though, it would be a manual inspection/removal/retry process to narrow down the content this extension cannot process.

If approaching this from a developer level, one could tweak the `visit_reference` implementation to dump the following:

```
def visit_reference(self, node):
    if self._reference_context:
        print('\nnested reference found')
        print('[docname]')
        print(self.docname)
        print('[current node]')
        print(node)
        print('[parent node]')
        print(node.parent)

    # nested references should never happen
    assert(not self._reference_context)
    
    ...
```
This may provide some useful information to continue an investigation for this issue.

**Comment by Tanoti** [comment]
I've done a bit of digging and I think it is because I have a custom role defined and then at two sections of the document structure the same target string is being generated using that role despite them actually being two different targets.

For example, in ``states/apache.rst`` I have:
```
.. sls:: apache
```
and in ``pillar/apache.rst`` I also have:
```
.. sls:: apache
```
Sphinx will render these properly and seems to point an incoming reference to either the closest target or whichever one was processed last (I have not dug fully in to this).

However, taking the logic that confluence is going to need unique references as it would for document titles, I am going to see if I can use a more granular role for my cross-references and if that will satisfy your extension.

**Comment by Tanoti** [comment]
After some more investigation, I have found the rst syntax that will generate the assert failure. Basically, it is because both a "contents" and a "doc" directive point to the same sub-document from one parent. Take this file set for example:

index.rst
```
Index
*******

.. toctree::
    :maxdepth: 1

    main
```

main.rst
```
main
*****

Elements
^^^^^^^^
.. contents::
    :depth: 1
    :local:

:doc:`common`
-------------
```

common.rst
```
:orphan:

common
******

Some content
```

Running the standard html builder on this generates the correct output but the confluence builder will fail with the the assert error in my original post when it tries to process main.rst.

Hope this helps!

**Comment by jdknight** [comment]
A solution (outlined in #287) has been applied to `master` and will be made available on PyPI next stable release (v1.3+).

Note that choice was made to ensure local table of contents headers are linked to locally, instead of how the HTML builder appears to link a local table entry directly to a header-referenced document. This is to be more in line to what reStructuredText's documentation states; and it appears to be a more sane approach (please mention otherwise). For the example outlined in this issue, the local table of contents will point to the local "common" heading on the page, and the heading on the page can be used to link to the other document.

**Comment by jdknight** [comment]
@Tanoti, thanks again for the example. With the merge into `master`, will go ahead and close this issue. If you have any other concerns or issues related to the changes, please do not hesitate to mention them.
